MSAL Node's Managed Identity Api Implementation

Robbie-Microsoft opened the pull request:
This PR covers the API for the new (MSAL-Node specific) Managed Identity Client. Only the App Service source has been implemented for now. The Azure Arc, CloudShell, IMDS and Service Fabric sources will be in another PR.

The following is how a developer will create a new managed identity client and acquire a token

```
import * as msal from "@azure/msal-node";

// optional
const config: msal.ManagedIdentityConfiguration = {
    // optional
    managedIdentityIdParams: {
        userAssignedClientId: "managed_identity_id"
    },
    system: { // optional
        loggerOptions: msal.LoggerOptions; // optional
        networkClient: msal.INetworkModule; // optional
        proxyUrl: string; // optional
        customAgentOptions: http.AgentOptions | https.AgentOption; // optional
    },
};

// required
const request: msal.ManagedIdentityRequest = {
    forceRefresh: true, // optional, defaults to false
    resource: "URI_of_the_managed_identity_assigned_to_the_azure_resource" // required
};

..................

try {
    const managedIdentityClient = new msal.ManagedIdentityClient(config);
    const response = await managedIdentityClient.acquireToken(request);
    console.log(response);
} catch (error) {
    console.log(error);
}
```

Robbie-Microsoft committed:
Implemented public api

bgavrilMS commented:
CC @pmaytak and @neha-bhargava on API proposal

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityClient.ts:
```diff
@@ -0,0 +1,157 @@
+/**
+ * Class to initialize a managed identity and identify the service
+ */
+export class ManagedIdentityClient {
```
Will this work though? For example are you going to be able to reuse the token caching logic that is already tied to ConfidentialClientApplication?

bgavrilMS commented on lib/msal-node/src/config/Configuration.ts:
```diff
@@ -107,6 +111,14 @@ export type Configuration = {
+export type ManagedIdentityConfiguration = {
+    forceRefresh?: boolean;
+    id?: string;
+    resourceType: ManagedIdentityIdType;
```
Not part of the public API. MSAL deals with this internally.

Robbie-Microsoft committed:
Implemented Feedback and made changes to public API

Robbie-Microsoft committed:
made ManagedIdentityConfiguration optional

Robbie-Microsoft committed:
made forceRefresh optional

Robbie-Microsoft committed:
ran prettier

neha-bhargava commented on lib/msal-node/src/client/ManagedIdentityApplication.ts:
```diff
@@ -0,0 +1,137 @@
+                    this.networkManager,
+                    this.managedIdentityId
+                ) ||
+                Imds.tryCreate(
```
For IMDS, we directly call the constructor rather than try create the source. If none of the above sources are available IMDS will be used by default.

Robbie-Microsoft replied:
I'm adding extra error checking in here. Maybe all the environment variables are malformed?

neha-bhargava commented on lib/msal-node/src/config/Configuration.ts:
```diff
@@ -107,6 +111,11 @@ export type Configuration = {
 
+export type ManagedIdentityConfiguration = Omit<
+    Configuration,
+    "auth" | "broker" | "cache" | "telemetry"
```
Does MSAL Node support any telemetry other than server side telemetry?

Robbie-Microsoft replied:
yes, it supports client telemetry. I will need to implement that eventually

bgavrilMS replied:
Actually I don't think you need to do that @Robbie-Microsoft . The way public client and SPA **client** telemetry is collected is different than confidential client. @neha-bhargava will do a demo on this soon I've heard :)

neha-bhargava commented on lib/msal-node/src/config/Configuration.ts:
```diff
@@ -199,3 +208,25 @@ export function buildAppConfiguration({
+}: ManagedIdentityConfiguration): ManagedIdentityNodeConfiguration {
+    const systemOptions: Required<NodeSystemOptions> = {
+        ...DEFAULT_SYSTEM_OPTIONS,
+        loggerOptions: system?.loggerOptions || DEFAULT_LOGGER_OPTIONS,
```
Not sure but is this where the AppTokenCache should be enabled and default authority should be set?

Robbie-Microsoft replied:
No, they should be initialized in ManagedIdentityApplication.ts

pmaytak commented on lib/msal-node/src/utils/Constants.ts:
```diff
@@ -3,6 +3,37 @@
 
+// MSI Constants. Docs for MSI are available here https://docs.microsoft.com/azure/app-service/overview-managed-identity
+export const APP_SERVICE_MSI_API_VERSION: string = "2019-08-01";
+export const SECRET_HEADER_NAME: string = "X-IDENTITY-HEADER";
```
The constants that are specific to each source should probably go into those specific files, so it's clearer to which source they apply. (see [MSAL.NET](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/9029e1c78aa62a918ce1db1aea54a2cfe3c15293/src/client/Microsoft.Identity.Client/ManagedIdentity/AppServiceManagedIdentitySource.cs#L19))

pmaytak commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,90 @@
+    ): ManagedIdentityId {
+        // TODO: throw exception if objectId is null
+        return new ManagedIdentityId(
+            ManagedIdentityIdType.USER_ASSIGNED_RESOURCE_ID,
```
```suggestion
            ManagedIdentityIdType.USER_ASSIGNED_OBJECT_ID,
```

Robbie-Microsoft replied:
good catch

pmaytak commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,90 @@
+    ): ManagedIdentityId {
+        // TODO: throw exception if clientId is null
+        return new ManagedIdentityId(
+            ManagedIdentityIdType.USER_ASSIGNED_RESOURCE_ID,
```
```suggestion
            ManagedIdentityIdType.USER_ASSIGNED_CLIENT_ID,
```

pmaytak commented on lib/msal-node/src/client/ManagedIdentityApplication.ts:
```diff
@@ -0,0 +1,137 @@
+/**
+ * Class to initialize a managed identity and identify the service
+ */
+export class ManagedIdentityApplication {
```
So do I understand the design:
- ManagedIdentityApplication is equivalent to ConfidentialClientApplication.
- Then each MI source class will be equivalent to ClientCredentialClient/OnBehalfOfClient/etc - classes that actually check the cache and get the tokens?

Right now, I do agree that inheriting from ClientApplication or BaseClient is not really useful.

bgavrilMS replied:
No, the "Sources" are an implementation detail. There is only 1 "AcquireToken" method. The user only needs to know: 

- their managed identity id (for user assigned) 
- the resource they want a token for (e.g. KeyVault)

That's it. No clientId, tenantid, authority. No cache either. 

Have a look at https://learn.microsoft.com/en-us/entra/msal/dotnet/advanced/managed-identity#examples for the .NET sample.

pmaytak reviewed (approved):
Yea, I think looks good so far, I think once at least 1 source is implemented, the full picture would be a bit clearer.

Robbie-Microsoft committed:
Updates to public api

Robbie-Microsoft committed:
Implemented non-public api

Robbie-Microsoft committed:
Code improvements + cleanup

Robbie-Microsoft committed:
added default cancellation token, and revised checks in httpClient

bgavrilMS commented on lib/msal-common/src/config/ClientConfiguration.ts:
```diff
@@ -7,7 +7,10 @@ import { INetworkModule } from "../network/INetworkModule";
-import { Constants } from "../utils/Constants";
+import {
+    Constants,
+    DEFAULT_TOKEN_RENEWAL_OFFSET_SEC,
```
Why do we need to add this now? It should just default to 5 min

bgavrilMS commented on lib/msal-common/src/network/INetworkModule.ts:
```diff
@@ -39,7 +39,8 @@ export interface INetworkModule {
         url: string,
-        options?: NetworkRequestOptions
+        options?: NetworkRequestOptions,
+        cancellationToken?: number
```
Do we really need to add timeout to this interface? If we really do, can we add it as part of `options` ?

Btw, if this is a timeout, just call it timeout, not "token". CancellationToken has more functionality, timeout is just one of them.

Robbie-Microsoft replied:
It seems to me that this was a bug where the timeout was missing from the post request definition in InNetworkModule. In httpClient, we allow the timeout (cancellationToken) to be optionally provided. @sameerag, can you chime in here? Was this originally supposed to be timeout functionality, or a cancellationToken?

bgavrilMS replied:
Let's not get bogged down in this @Robbie-Microsoft - just open a bug and we'll treat it separately. Let's not make changes to the existing interface if they are not needed.

Robbie-Microsoft replied:
Well regardless, this change is needed for Managed Identity if we're exposing the timeout to the developer.

But per your other comment, we can start without exposing this value and always add it later.

bgavrilMS commented on lib/msal-common/src/config/ClientConfiguration.ts:
```diff
@@ -18,9 +21,6 @@ import { ISerializableTokenCache } from "../cache/interface/ISerializableTokenCa
 import { ProtocolMode } from "../authority/ProtocolMode";
 
-// Token renewal offset default in seconds
-const DEFAULT_TOKEN_RENEWAL_OFFSET_SEC = 300;
```
Is this part of the public API? Isn't this a breaking change?

Robbie-Microsoft replied:
It is part of the public api, and it's not a breaking change. I updated the imports in msal-common.

It's used in all clients (public api, msal-common), but Managed Identity (msal-node) needs this constant.

bgavrilMS commented on lib/msal-common/src/network/NetworkManager.ts:
```diff
@@ -43,23 +43,62 @@ export class NetworkManager {
+     * @param tokenEndpoint
+     * @param options
+     */
+    async sendGetRequest<T extends ServerAuthorizationTokenResponse>(
```
We discussed that `NetworkManager` is actually `TokenEndpointNetworkManager`. But in MSI case, you'll be hitting different endpoints, which will not return `ServerAuthorizationTokenResponse`. So I don't think `NetworkManager` can be used here. 

Consider creating an `MsiNetworkManager` or using the `networkClient` directly.

Robbie-Microsoft replied:
Based on your other comment about MSI not needing throttling, I think it makes more sense to just call the networkClient directly.

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityApplication.ts:
```diff
@@ -0,0 +1,275 @@
+        | AzureArc
+        | Imds;
+
+    private fakeAuthority: Authority;
```
Pls add a comment here on why this is fake etc.

bgavrilMS commented on lib/msal-node/src/config/Configuration.ts:
```diff
@@ -83,6 +84,8 @@ export type NodeSystemOptions = {
     networkClient?: INetworkModule;
     proxyUrl?: string;
     customAgentOptions?: http.AgentOptions | https.AgentOptions;
+    tokenRenewalOffsetSeconds?: number;
```
This is not a system option. Why do we need this to be configurable anyway?

Robbie-Microsoft replied:
It's used to compare the access token to, to see if the access token is expired. This is something that already exists in CommonClientConfiguration, but MSI does not extend CommonClientConfiguration. We could just keep it at it's default value of 300 ms and not allow the developer to optionally configure it, if you'd like.

Robbie-Microsoft replied:
I believe it is a system option. Look in lib/msal-node/src/client/ClientCredentialClient.ts on line 127

bgavrilMS replied:
Can we not expose this for MSI please? I have no idea why it's exposed for ConfidentialClient, there are no scenarios that I know of where we'd want app developers to change this. Maybe in Public Client and SPA, but not in Confidential Client and MSI.

Robbie-Microsoft replied:
How should we check if the access token is expired? Use the default value?

bgavrilMS replied:
Yes. 5 min.

bgavrilMS commented on lib/msal-node/src/config/Configuration.ts:
```diff
@@ -83,6 +84,8 @@ export type NodeSystemOptions = {
     proxyUrl?: string;
     customAgentOptions?: http.AgentOptions | https.AgentOptions;
+    tokenRenewalOffsetSeconds?: number;
+    cancellationToken?: number;
```
We should not add this new option just for MSI. Cancellation / timeout always existed in MSAL .NET but we are not bringing it in NodeJS as part of MSI. If ppl need it, we can bring it separately. 

Cancellation of async operations in NodeJS is different than cancellation in .NET

Robbie-Microsoft replied:
I mentioned this in a response to the previous comment, but these already exist in the CommonClientConfiguration which all other msal-node clients extend. MSI doesn't extend CommonClientConfiguration, so I added them to the MSI configuration. Thoughts? @sameerag what are your thoughts?

bgavrilMS replied:
Let's start without. We can always add later?

bgavrilMS commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,98 @@
+                this._isUserAssignedId = true;
+                break;
+            default:
+                // TODO: throw error
```
There are a few simple TODOs here that can be easily solved.

bgavrilMS commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,98 @@
+    /**
+     * Creates an instance of ManagedIdentityId for a system assigned managed identity
+     */
+    public static createSystemAssigned(): ManagedIdentityId {
```
How about `public static get systemAssignedIdentity()`

Robbie-Microsoft replied:
resolved per our conversation

bgavrilMS commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,98 @@
+ */
+
+import {
+    DEFAULT_MANAGED_IDENTITY_ID,
```
Is this constant part of the public API? Why?

Robbie-Microsoft replied:
I don't understand your question.

Robbie-Microsoft replied:
resolved per our conversation

bgavrilMS commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,98 @@
+} from "../utils/Constants";
+
+export class ManagedIdentityId {
+    private _id: string;
```
inconsistent naming. Pls don't follow the .NET naming conventions. There is a linter in MSAL JS, pls ensure it runs and passes.

Robbie-Microsoft replied:
I don't understand what you're saying. I run the linter before every push to GitHub.

bgavrilMS replied:
If I look at existing files like `ClientCredentialClient.ts` I see that private variables are defined as 

`private scopeSet: ScopeSet;`

But in your case you use 

`private _id: string`

So the naming is inconsistent. Maybe the linter doesn't work or doesn't pick up on this? In any case, pls use consistent naming of vars.

bgavrilMS commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,98 @@
+    ManagedIdentityIdType,
+} from "../utils/Constants";
+
+export class ManagedIdentityId {
```
Since this is JS and ppl like JSON, I think you should have an object model that is more declarative and less static-y. 
```json
{ 
    "systemAssigned": true
}
```
```json
{ 
    "systemAssigned": false,
    "userAssignedClientId": 12345
}
```

```json
{ 
    "systemAssigned": false,
    "userAssignedObjectId": 12345
}
```

bgavrilMS replied:
@rayluo for a perspective, as I think this is what we did in MSAL PY as well.

Robbie-Microsoft replied:
I've started implementing this. I think this is more error prone as the developer could input all three of userAssignedClientId, userAssignedResourceId and userAssignedObjectId. I can add if statements to mitigate this, but it will be much more code than before.

bgavrilMS replied:
Ok, how about this then. Let's support the JSON approach for the most common use case. I believe 99% of app developers rely on system assigned identity or on user assigned identity + client_id. 

```json
{ 
    "systemAssigned": true
}
```

```json
{ 
    "systemAssigned": false, // optional, but if set, it can't be true
    "userAssignedIdenityClientId": 12345
}
```

The 1% needing user assigned identity by resourceId can use the static method.

pmaytak replied:
So @bgavrilMS you're suggesting not to follow how .NET did it? +1 to Robbie's comment that I would narrow the API down. We could allow the user to set (ManagedIdentityIdType, UserProvidedId?) (although in .NET IdType is internal).

bgavrilMS replied:
MSAL .NET is not JSON driven, but MSAL Node is and Id.Web are. So proposal is more like: https://github.com/AzureAD/microsoft-identity-web/issues/1775#issuecomment-1769864869

Agree on hiding "IdType" - this is implementation detail. 

@Robbie-Microsoft - can you please organize a meeting and include all reviewers and @localden and discuss the public API you have ? In particular we need to see how to specify SAMI and UAMI.

Robbie-Microsoft replied:
I've implemented this

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityApplication.ts:
```diff
@@ -0,0 +1,275 @@
+
+    private fakeAuthority: Authority;
+
+    constructor(
```
For consistency with `ConfidentialClientApplication`, we should have only 1 config param here, not 2. And that config param should be mandatory.

Robbie-Microsoft replied:
Making this change would mean editing the MSI config and making managedIdentityId a required part of the config, provided by the developer. I can do that if you'd like.

bgavrilMS replied:
I think that's a better developer experience? Managed Identity ID is mandatory.

This is very similar to `client_id`,. which is mandatory for CCA

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityApplication.ts:
```diff
@@ -0,0 +1,275 @@
+
+        this.cryptoProvider = new CryptoProvider();
+
+        this.identitySource = this.selectManagedIdentitySource();
```
We shouldn't try to select the MI source when creating the app, but when firing the request.
Once a request has figured out what MI source we're on (e.g. VM), all other requests need to remeber this (i.e. static cache this)

Robbie-Microsoft replied:
I disagree. If we select the MSI source when creating the app, then it will already be known when firing the requests. Isn't this how .dotnet does it?

bgavrilMS replied:
No, .NET selects the source when making the request itself. It then caches the sources in a **static** variable (so for the duration of the process). 

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/Internal/Requests/ManagedIdentityAuthRequest.cs#L156

The important bit here is to cache it statically, so that no matter how many MSIAppcalition objects you created, the discovery process should only occur once. 
For consistency purposes, I'd keep this at request level.

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityApplication.ts:
```diff
@@ -0,0 +1,275 @@
+            version
+        );
+
+        this.cacheManager = new DefaultStorageClass(
```
What is the difference between `NodeStorage`, which ClientApplicaiton uses and this one?

Robbie-Microsoft replied:
Good question, I'll have to look into this more. It appears NodeStorage is used in TokenCache, and TokenCache's methods are exposed to the user. For example: serialize, deserialize, getAllAccounts, etc. I don't think we want to expose these to the developer of an MSI application? If not, CacheManager should suffice for storing tokens in a cache. @sameerag am I understanding this correctly?

Robbie-Microsoft replied:
On second pass, it looks like NodeStorage extends CacheManager, which this uses. 

NodeStorage "implements Storage for node, reading cache from user specified storage location or an  extension library". Managed Identity doesn't do either of those thing, so CacheManager should suffice.

Robbie-Microsoft replied:
resolved per our discussion

Robbie-Microsoft replied:
Another Update. I made a mistake when first using DefaultStorageClass. None of the cache methods are implemented in DefaultStorageClass. I should have been using NodeStorage initially.

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityApplication.ts:
```diff
@@ -0,0 +1,275 @@
+        );
+
+        this.networkClient = this.config.system.networkClient;
+        this.networkManager = new NetworkManager(
```
Forget about network manager. It is a wrapper over networkClient that adds throttling. The throttling logic is specific to talking to ESTS and does not apply here.

bgavrilMS commented on lib/msal-node/src/request/ManagedIdentityRequest.ts:
```diff
@@ -0,0 +1,16 @@
+import { BaseAuthRequest } from "@azure/msal-common/request/BaseAuthRequest";
+
+/**
+ * ManagedIdentityRequest
```
Please add better comments, based on MSAL.NET ones which were already reviewed etc.

bgavrilMS commented on lib/msal-node/src/request/ManagedIdentityRequest.ts:
```diff
@@ -0,0 +1,16 @@
+/**
+ * ManagedIdentityRequest
+ * - forceRefresh - forces managed identity requests to skip the cache and make network calls if true
+ * - resourceUri  - the URI of the managed identity
```
See https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/IManagedIdentityApplication.cs 

Please add a task to the board to make sure we support both formats. Can be addressed separately. For now just focus on "resourceUri"

Robbie-Microsoft replied:
I don't understand? forceRefresh is optional here.

bgavrilMS replied:
I am not talking about force refresh, I am talking about the format of the resource URI string (btw, please rename the param as `resource` since it's not a URI ). 

See https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/IManagedIdentityApplication.cs#L22

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityApplication.ts:
```diff
@@ -0,0 +1,275 @@
+        managedIdentityId: ManagedIdentityId
+    ): Promise<AuthenticationResult | null> {
+        return (
+            (await this.getCachedToken(
```
Think how you can refactor the code from ClientCredentialClient so as to reuse the fetching from cache logic. We should not rewrite this because it is  common between client_credentials flow and MSI. For example, you already have a bug in your code because you don't skip the cache when forceRefresh=true. 

I see 2 ways: 

- create a ClietnCredentialsRequest and set an AppTokenProvider. The AppTokenProvider makes MSI calls etc. Caching is ensured by the ClietnCredentialsRequest.
- OR share code from ClietnCredentialsRequest 

I'd go for the first one, but @neha-bhargava can also suggest.

Robbie-Microsoft replied:
I don't understand this comment. Refactoring the code from ClientCredentialClient is exactly how I wrote the caching logic here. I do skip the cache when forceRefresh is true... on line 144.

bgavrilMS replied:
ClientCredentialClient does more things like calculates the outcome and does pro-active token refresh. We need that functionality as well. See `getCachedAuthenticationResult` - can we reuse that?

Robbie-Microsoft replied:
That is essentially what I've already done, but I left those parts out. I will add them.

neha-bhargava replied:
I think we should reuse the code for getting token from cache and saving the token to cache from client credential. And add conditions to skip the parts not relevant to managed identity.

Robbie-Microsoft committed:
fixed build errors

Robbie-Microsoft committed:
Undid changes to network manager. calling network client directly now

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentityApplication.ts:
```diff
@@ -0,0 +1,270 @@
+    public async acquireToken(
+        managedIdentityRequest: ManagedIdentityRequest,
+        managedIdentityId: ManagedIdentityId
+    ): Promise<AuthenticationResult | null> {
```
Will it ever return null? I don't think so. It should always return `AuthenticationResult` or throw exception

Robbie-Microsoft replied:
It will return null if forceRefresh has been set to true, if there is no cached access token, or if the cached access token is expired. This is how the ClientCredentials flow functions as well. I think it makes more sense to return null.

Robbie-Microsoft replied:
resolved, per our conversation

Robbie-Microsoft committed:
Implemented some github feedback

Robbie-Microsoft committed:
Implemented some GitHub feedback

Robbie-Microsoft committed:
Implemented GitHub Feedback

Robbie-Microsoft committed:
undefined managed identity config means the managed identity is system-assigned

Robbie-Microsoft committed:
fixed broken unit tests

Robbie-Microsoft committed:
Deleted msal-node's package-lock.json (#6571)

This file should have been deleted as part of the original npm
workspaces PR. It breaks npm workspaces when developing in msal-node.

rgins16 committed:
Merge branch 'msi_feature_branch' into public_api

rgins16 committed:
Merge branch 'msi_feature_branch' into public_api

rgins16 committed:
Added unit tests. Fixed bugs and made improvements based on unit tests.

bgavrilMS commented on lib/msal-common/src/cache/CacheManager.ts:
```diff
@@ -1432,11 +1432,16 @@ export abstract class CacheManager implements ICacheManager {
-                staticAliases.includes(entity.environment)
-            ) {
+
+            const validEnvironment = staticAliases?.some((staticAlias) => {
```
What was the bug? Is it related to the use "indexOf" instead of "includes" ?

pmaytak replied:
Why doesn't `includes` work here?
Is this related to managed identity? If not, I think it should be included in a separate PR into dev.

Robbie-Microsoft replied:
These are the values during Managed Identity Flow:
```
staticAlias = [ 'https://login.microsoftonline.com/common/' ]
environment:  login.microsoftonline.com
entity.environment:  https://login.microsoftonline.com/common/
```

The existing logic did not recognize that there is indeed a match here. The authority is fake, and so is the environment, which is why there is an issue here.

pmaytak commented on lib/msal-common/src/utils/Constants.ts:
```diff
@@ -442,3 +442,6 @@ export const JsonTypes = {
 export const ONE_DAY_IN_MS = 86400000;
+
+// Token renewal offset default in seconds
+export const DEFAULT_TOKEN_RENEWAL_OFFSET_SEC = 300;
```
If this is not related to managed identity, separate into a separate PR?

pmaytak commented on lib/msal-common/src/response/AuthenticationResult.ts:
```diff
@@ -44,4 +44,6 @@ export type AuthenticationResult = {
     msGraphHost?: string;
     code?: string;
     fromNativeBroker?: boolean;
+    managedIdentityClientId?: string;
+    managedIdentityResource?: string;
```
Do we need to return these? The user already knows what they are since they pass them into a request.

neha-bhargava replied:
We do not need the managedIdentityClientId but pass the managedIdentityResource as scopes on line 29.

bgavrilMS replied:
Also this is a public type, any changes to public API should be highlighted and reviewed. 

In this case, the change is not needed. AuthResult does not need to change for MSI.

pmaytak commented on lib/msal-node/src/config/Configuration.ts:
```diff
@@ -107,6 +108,17 @@ export type Configuration = {
     telemetry?: NodeTelemetryOptions;
 };
 
+export type ManagedIdentityIdParams = {
+    userAssignedClientId?: string;
+    userAssignedResourceId?: string;
+    userAssignedObjectId?: string;
+};
+
+export type ManagedIdentityConfiguration = {
+    managedIdentityIdParams?: ManagedIdentityIdParams;
+    system?: NodeSystemOptions;
+};
```
What about instead of `ManagedIdentityIdParams`, `ManagedIdentityConfiguration` accepts `ManagedIdentityIdType` (required) and `UserAssignedId` (optional, in case of system assigned Id) to make API clearer. Currently with `ManagedIdentityIdParams`, what if multiple IDs are set, what if none are set?

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/AppConfig/ManagedIdentityId.cs

pmaytak commented on lib/msal-node/src/config/Configuration.ts:
```diff
@@ -199,3 +211,31 @@ export function buildAppConfiguration({
+    system: Required<NodeSystemOptions>;
+};
+
+export function buildManagedIdentityConfiguration({
+    managedIdentityIdParams,
+    system,
+}: ManagedIdentityConfiguration): ManagedIdentityNodeConfiguration {
```
```typescript
({
    managedIdentityIdParams,
    system,
}: ManagedIdentityConfiguration)
```
What does this syntax mean? This method accepts an instance of `ManagedIdentityConfiguration`, and it's syntactic sugar, so you don't have to call config.paramBlah?

Robbie-Microsoft replied:
It's called [Destructuring Assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment). Scroll down to the example with function parameters, like what you're looking at here.

pmaytak commented on lib/msal-node/src/config/Configuration.ts:
```diff
@@ -107,6 +108,17 @@ export type Configuration = {
+    userAssignedObjectId?: string;
+};
+
+export type ManagedIdentityConfiguration = {
```
What about instead of `ManagedIdentityIdParams`, `ManagedIdentityConfiguration` accepts `ManagedIdentityIdType` (required) and `UserAssignedId` (optional, in case of system assigned Id) to make API clearer. Currently with `ManagedIdentityIdParams`, what if multiple IDs are set, what if none are set?

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/AppConfig/ManagedIdentityId.cs

bgavrilMS replied:
Are you talking about the public API here @pmaytak ? I think the public API in NodeJS should be JSON driven and similar to the config option (also JSON) from [ID.Web](https://github.com/AzureAD/microsoft-identity-web/issues/1775)

<img width="559" alt="image" src="https://github.com/AzureAD/microsoft-authentication-library-for-js/assets/12273384/712fa2ea-0751-41c2-99f6-5e1117ad72f5">

pmaytak replied:
I didn't understand what you meant by "JSON-driven", makes sense it's about config now. Yea, I think it's fine to do a check and throw if multiple mutually exclusive IDs are set. The other option of specifying Type and IdValue is probably less intuitive.

Robbie-Microsoft replied:
I've implemented this

pmaytak commented on lib/msal-node/src/config/Configuration.ts:
```diff
@@ -199,3 +211,31 @@ export function buildAppConfiguration({
     };
 }
+
+export type ManagedIdentityNodeConfiguration = {
```
Can `ManagedIdentityNodeConfiguration` and `ManagedIdentityConfiguration` be combined? Let the user provide `ManagedIdentityId`? These classes look very similar.

Robbie-Microsoft replied:
I based this off existing implementation for Configuration and NodeConfiguration (for existing clients). In this case, Configuration is passed in by the user, and NodeConfiguration is what's sent to the client. Optional parameters not provided by the developers' Configuration are set to default values here.

pmaytak commented on lib/msal-node/src/config/ManagedIdentityRequestParameters.ts:
```diff
@@ -0,0 +1,54 @@
+        );
+    }
+
+    public computeParametersBodyString(): string {
```
What is this meant to do?

Robbie-Microsoft replied:
It does a pre-transformation to a network request's body; it turns it into a string. In the case of managed identity, I think forceRefresh will be the only body parameter that could ever be sent.

pmaytak commented on lib/msal-node/src/request/ManagedIdentityRequest.ts:
```diff
@@ -0,0 +1,15 @@
+ * - forceRefresh - forces managed identity requests to skip the cache and make network calls if true
+ * - resource  - resource requested to access the protected API. It should be of the form "{ResourceIdUri}" or {ResourceIdUri/.default}. For instance https://management.azure.net or, for Microsoft Graph, https://graph.microsoft.com/.default
+ */
+export type ManagedIdentityRequest = ManagedIdentityRequestParams &
```
I assume this syntax means `ManagedIdentityRequest` inherits all properties from both of these classes?

Robbie-Microsoft replied:
yep

pmaytak commented on lib/msal-node/src/request/ManagedIdentityRequestParams.ts:
```diff
@@ -0,0 +1,14 @@
+ * - forceRefresh - forces managed identity requests to skip the cache and make network calls if true
+ * - resource  - resource requested to access the protected API. It should be of the form "{ResourceIdUri}" or {ResourceIdUri/.default}. For instance https://management.azure.net or, for Microsoft Graph, https://graph.microsoft.com/.default
+ */
+export type ManagedIdentityRequestParams = {
```
- There's already similarly named `ManagedIdentityRequestParameters` class. Can this class be eliminated and these properties directly be in `ManagedIdentityRequest`, and then for consistency `ManagedIdentityApplication.acquireToken` will accept `ManagedIdentityRequest` instance?

Robbie-Microsoft replied:
I see your confusion. However, ManagedIdentityRequest only exists so that caching functionality from ClientCredentials can be re-used by ManagedIdentity. This functionality requires an authority, correlationId, etc. The  developer can't pass in this information, they can only pass in forceRefresh and resource, which are in ManagedIdentityRequestParameters.

pmaytak commented on lib/msal-node/src/request/ManagedIdentityRequestParams.ts:
```diff
@@ -0,0 +1,14 @@
+ * - resource  - resource requested to access the protected API. It should be of the form "{ResourceIdUri}" or {ResourceIdUri/.default}. For instance https://management.azure.net or, for Microsoft Graph, https://graph.microsoft.com/.default
+ */
+export type ManagedIdentityRequestParams = {
+    forceRefresh?: boolean;
+    resource: string;
```
Does JS care about order of required and optional parameters or is there some coding convention? In .NET required parameters must come before any optional ones.

Robbie-Microsoft replied:
It doesn't matter for a type definition, which this is. In this case, I put them in alphabetical order.

pmaytak commented on lib/msal-common/src/response/ServerAuthorizationTokenResponse.ts:
```diff
@@ -51,4 +51,15 @@ export type ServerAuthorizationTokenResponse = {
+     * Unique to Managed Identity
+     */
+    // success
+    client_id?: string;
```
Since these properties are kind of separate from the above ones, can they be extracted into something like `ManagedIdentityResponse` and then called like `networkClient.sendPostRequestAsync<ManagedIdentityResponse>?

Robbie-Microsoft replied:
I could, and that's how I originally did it, but then I'd have to change more code in msal-common, which is what I'm trying to avoid. I'd have two different server responses. In ResponseHandler on line 399, I'd have to change the type from ServerAuthorizationTokenResponse to ServerAuthorizationTokenResponse | ManagedIdentityTokenResponse, and then the rest of the code in that function would throw errors and I'd have to add a bunch of conditional checks to see if properties on the server response exist. Thoughts?

bgavrilMS replied:
Agree with Peter, this is violating the Single Responsability Principle. This object is only for ESTS responses. We should have a different response for MSI - MSIResponse and if needed, an adapter (e.g. method on ServerAuthorizationResponse) that converts between MSIResponse to ServerAuthorizationResponse.

pmaytak commented on lib/msal-node/src/utils/Constants.ts:
```diff
@@ -3,6 +3,44 @@
+/**
+ * Managed Identity Authentication Sources
+ */
+export const ManagedIdentitySource = {
```
Is this used anywhere?

Robbie-Microsoft replied:
no, I think I can delete this

pmaytak commented on lib/msal-node/src/client/ManagedIdentityApplication.ts:
```diff
@@ -0,0 +1,179 @@
+     */
+    public async acquireToken(
+        managedIdentityRequestParams: ManagedIdentityRequestParams
+    ): Promise<ManagedIdentityResult> {
```
Shouldn't this return `AuthenticationResult` like other APIs?

Robbie-Microsoft replied:
No, a managed identity result has been created which will be returned to the developer.

pmaytak commented on lib/msal-node/src/client/ManagedIdentityClient.ts:
```diff
@@ -0,0 +1,126 @@
+                );
+            }
+            return source;
+        } catch (error) {
```
nit: try/catch not needed if we're just throwing?

pmaytak commented on lib/msal-node/src/client/ManagedIdentityClient.ts:
```diff
@@ -0,0 +1,126 @@
+        }
+    }
+
+    public async sendMSITokenRequest(
```
nit: `sendManagedIdentityTokenRequest` for consistency.

Robbie-Microsoft replied:
consistency with what?

pmaytak commented on lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts:
```diff
@@ -0,0 +1,141 @@
+        managedIdentityId: ManagedIdentityId
+    ): ManagedIdentityRequestParameters;
+
+    public async authenticateWithMSI(
```
nit: `acquireTokenWithManagedIdentity` or `authenticateWithManagedIdentity`

pmaytak commented on lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts:
```diff
@@ -0,0 +1,141 @@
+export abstract class BaseManagedIdentitySource {
+    protected logger: Logger;
+    private cacheManager: CacheManager;
+    private networkClient: INetworkModule;
```
I couldn't find it but what is the actual implementation of INetworkMobile that gets passed in here?

Robbie-Microsoft replied:
It's an interface with two methods that aren't implemented: sendGetRequestAsync and sendPostRequestAsync. It's in msal-common: lib/msal-common/src/network/INetworkModule.ts

rgins16 committed:
Small improvements

Robbie-Microsoft committed:
Implemented IMDS

Robbie-Microsoft committed:
small improvements

Robbie-Microsoft committed:
bugfixes

Robbie-Microsoft committed:
Modification to HttpClient to account for Managed Identity always sending http (not https) requests

Robbie-Microsoft committed:
Small improvement to core functionality and unit tests

Robbie-Microsoft committed:
Added IMDS unit tests and made small changes to core functionality for unit tests purposes

neha-bhargava commented on lib/msal-common/src/cache/entities/AccessTokenEntity.ts:
```diff
@@ -76,7 +78,9 @@ export class AccessTokenEntity extends CredentialEntity {
-        requestedClaimsHash?: string
+        requestedClaimsHash?: string,
+        managedIdentityClientId?: string,
+        managedIdentityResource?: string
```
I think you can avoid adding these 2 parameters and just pass managedIdentityClientId as clientId and managedIdentityResource as scopes. Should work fine to generate cache entity.

neha-bhargava commented on lib/msal-common/src/response/ResponseHandler.ts:
```diff
@@ -635,6 +640,12 @@ export class ResponseHandler {
                 cacheRecord.account?.msGraphHost || Constants.EMPTY_STRING,
             code: serverTokenResponse?.spa_code,
             fromNativeBroker: false,
+            managedIdentityClientId:
```
Does not seem right to have managedIdentityClientId and managedIdentityResource added in general. These parameters map to clientId and scopes respectively. Any method that has scopes in the flow should ideally contain the resource from managed identity request and client id should be managedIdentityClientId.

bgavrilMS commented on lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    // used in unit tests
+    public isAppService(): boolean {
```
This is a code smell. Base class should not know about derived class. 

Why don't you just copy this logic to the tests?

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+                        .getIsUserAssignedId
+                ).toEqual(false);
+
+                // wait two seconds for the token to be saved to the cache, before checking the cache for the token
```
no dead code pls

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+                ManagedIdentityTestUtils.getManagedIdentityNetworkClient(
+                    DEFAULT_MANAGED_IDENTITY_ID
+                ),
+            // managedIdentityIdParams will be omitted for system-assigned
```
@localden  @pmaytak  - this (and line 43) is the public API currently. Please have a look.

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+        resource: MANAGED_IDENTITY_RESOURCE,
+    };
+
+    const nonSystemAssignedConfig: ManagedIdentityConfiguration = {
```
Please call this "userAssignedMSIConfig" or smth similar.

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+    describe("Acquires a token successfully via an App Service Managed Identity", () => {
+        beforeAll(() => {
+            // make a copy
+            process.env = { ...OLD_ENVS };
```
Isn't this done already for all tests?

Robbie-Microsoft replied:
no, the one at the top of the file defines OLD_ENVS

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+            process.env = OLD_ENVS;
+        });
+
+        test("acquires a Non-System-Assigned token", async () => {
```
Pls avoid calling it "non-system-assigned" and use the term "user assigned" when you mean user assigned + client ID and "user assigned by object ID" etc.

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+            const managedIdentityApplication = new ManagedIdentityApplication(
+                nonSystemAssignedConfig
+            );
+            expect(managedIdentityApplication.isAppService()).toBe(true);
```
Not a public API we want to support.

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+            );
+            expect(managedIdentityApplication.isAppService()).toBe(true);
+
+            const networkManagedIdentityResult =
```
What is the type here? `AuthenticationResult` ?

Robbie-Microsoft replied:
ManagedIdentityResult. I've updated the code to include the return type

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+                DEFAULT_MANAGED_IDENTITY_AUTHENTICATION_RESULT.accessToken
+            );
+            expect(
+                managedIdentityApplication.getConfig.managedIdentityId.getId
```
Is this null for System assigned?

Robbie-Microsoft replied:
no, it defaults to "system_assigned_managed_identity", the same way .net does

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+                managedIdentityApplication.getConfig.managedIdentityId.getId
+            ).toEqual(MANAGED_IDENTITY_CLIENT_ID);
+            expect(
+                managedIdentityApplication.getConfig.managedIdentityId.getIdType
```
Can we hide "getIdType" from the public API please?

Robbie-Microsoft replied:
see other comment

bgavrilMS commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,141 @@
+        return this.id;
+    }
+
+    public get getIdType(): ManagedIdentityIdType {
```
void using "getX" in getters. Use just "X". E.g. `public get Id`

bgavrilMS commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,141 @@
+        return new ManagedIdentityId();
+    }
+
+    // /**
```
Pls remove dead code from PRs

bgavrilMS commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,141 @@
+        return this.id;
+    }
+
+    public get getIdType(): ManagedIdentityIdType {
```
I don't think this should be public API

Robbie-Microsoft replied:
It needs to be declared public because the type is checked in AppService.ts and IMDS.ts. However, I can delete getIsUserAssigned because it's only used in the unit tests.

bgavrilMS replied:
I don't understand this. We fully control the public API?

bgavrilMS replied:
Could we have `ManagedIdentityId` be only a thin wrapper class for the config JSON and we keep all the extra details such as the IDType oject non-exportable?




```typescript

// This is the public API
export class ManagedIdentityId
{
      public UserAssignedClientId;
      public UserAssignedObjectId;
      public UserAssignedResourceId;
}

// Class that we use in MSAL interally only
class ManagedIdentityIdInternal {

     public readonly type: ManagedIdType 
     constructor(ManagedIdentityId) { 
          type = ... // figure out the type and throw exceptions if incorrect type specified
     }
```     

}

```

Robbie-Microsoft replied:
Even though this class is "exported", it's not available to the developer. In order for an msal-node class or type to be exposed to the developer, it must be exported in lib/msal-node/src/index.ts. In the change-files for this PR, you'll see that only ManagedIdentityApplication and ManagedIdentityRequestParams are exposed (exported).

bgavrilMS commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,141 @@
+        return this.idType;
+    }
+
+    public get getIsUserAssignedId(): boolean {
```
I don't think this should be public API

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+                    MANAGED_IDENTITY_CLIENT_ID
+                ),
+        },
+        managedIdentityIdParams: {
```
In hindsight, I think this would be better named as "managedIdentityConfig"

Robbie-Microsoft replied:
I don't think this makes sense because the type is already ManagedIdentityConfiguration

bgavrilMS replied:
Let's review the public API in a meeting.

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+            });
+        });
+    });
+});
```
We need tests that validate the access token cache key for all configurations - system assigned, user assigned + client etc.

neha-bhargava replied:
Updated the api review doc to add cache keys for all the managed identity scenarios. https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/ManagedIdentity/ManagedIdentity.md&version=GBdev&_a=preview&anchor=acceptance-tests-for-cache-keys.

Robbie-Microsoft replied:
I just read Neha's addition to the API review doc regarding the cache keys. I don't think I can write unit tests to verify them... the nodeStorage variable is private and can't be accessed from the unit tests. In Client Credential, we can pass in a custom node storage, and verify cache items that way. We've disabled that for Managed Identity, so I don't think I will be able to have access to the cache to check the cache key.

bgavrilMS commented on lib/msal-node/src/config/ManagedIdentityId.ts:
```diff
@@ -0,0 +1,141 @@
+        const userAssignedClientId =
+            managedIdentityIdParams?.userAssignedClientId;
+        const userAssignedResourceId =
+            managedIdentityIdParams?.userAssignedResourceId;
```
Are these 3 options all implemented for IMDS and MSI? If they are, are you missing tests?

Robbie-Microsoft replied:
I believe the implementation for each of them is one and the same. I need to write unit tests for each one though.

bgavrilMS commented on lib/msal-node/test/client/ManagedIdentityApplication.spec.ts:
```diff
@@ -0,0 +1,326 @@
+                        .getIsUserAssignedId
+                ).toEqual(false);
+            });
+        });
```
Please add more complex tests, for example: 

1. Request token for resource R1 for SAMI
2. Request token for resource R2 for SAMI
3. Request token for resource R2 for UAMI

Assert how they are in the cache (i.e. 3 separate tokens)

4. Make requests again and assert everything comes from the cache.

These tests do not need to be written for each individual source, as they mostly test the state of the cache.

@neha-bhargava - can you please articulate some more complex unit tests based on your work in MSAL Java  and MSAL .NET?

neha-bhargava replied:
Shared the tests in MSAL .Net for the scenarios: [microsoft-authentication-library-for-dotnet/tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/ManagedIdentityTests.cs at main · AzureAD/microsoft-authentication-library-for-dotnet (github.com)](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/ManagedIdentityTests.cs).

You can also look at some tests in MSAL Java: https://github.com/AzureAD/microsoft-authentication-library-for-java/pull/712/files#diff-c1e78d16db8b6cb86ef4e37c4f14c7950221b0fd619b2ab501499c588b5c0833.

One note page that contains some raw responses from the server can help with mocking: https://microsoft.sharepoint.com/teams/ADAL/_layouts/OneNote.aspx?id=%2Fteams%2FADAL%2FSiteAssets%2FDevEx%20Notebook&wd=target%28ID4S%2FMSALs%2FTechnical%2FMSI.one%7CA2A57687-9A17-4C9E-86D1-60AA32AB5005%2FMSI%20Errors%20%26%20Exceptions%7C41FE3612-4FAC-475D-9667-4A114AEBEEDA%2F%29

Robbie-Microsoft replied:
@bgavrilMS, I'm unable to implement that test you suggested. It requires two different managed identity applications, with two different mocked network responses. However, this is impossible, because the Managed Identity Application's ManagedIdentityClient.identitySource is static. Once created, we don't create it again (your feedback, and this means that it persists across Managed Identity Application instances). So whatever the mocked network response is upon creation... it will be unable to be changed for a different resource (client id).

Robbie-Microsoft replied:
If I comment out the following if statement, then the test I wrote works (based on what you listed above)
```
// if (!ManagedIdentityClient.identitySource) {
        ManagedIdentityClient.identitySource = this.selectManagedIdentitySource(
            this.logger,
            this.nodeStorage,
            this.networkClient,
            this.cryptoProvider
        );
        // }
```

bgavrilMS commented on lib/msal-node/test/test_kit/StringConstants.ts:
```diff
@@ -348,6 +350,22 @@ export const AUTHENTICATION_RESULT_DEFAULT_SCOPES = {
+    {
+        accessToken: TEST_TOKENS.ACCESS_TOKEN,
+        expiresOn: TEST_TOKEN_LIFETIMES.DEFAULT_EXPIRES_IN,
+        resource: MANAGED_IDENTITY_RESOURCE,
```
Is "resource" new? I thought AuthResult only has scopes.

Robbie-Microsoft replied:
not returning resource anymore since the developer already knows it

bgavrilMS commented on lib/msal-node/test/test_kit/StringConstants.ts:
```diff
@@ -348,6 +350,22 @@ export const AUTHENTICATION_RESULT_DEFAULT_SCOPES = {
+export const DEFAULT_MANAGED_IDENTITY_AUTHENTICATION_RESULT: ManagedIdentityResult =
+    {
+        accessToken: TEST_TOKENS.ACCESS_TOKEN,
+        expiresOn: TEST_TOKEN_LIFETIMES.DEFAULT_EXPIRES_IN,
```
This must be a bug. `expiresOn` is a date but expires_in is a number (of seconds)

Robbie-Microsoft replied:
I believe expiresOn should be a number. It's returned from the managed identity as a number. See [here](https://learn.microsoft.com/en-us/azure/app-service/overview-managed-identity?tabs=portal%2Chttp). We also return it as a number in lib/msal-node/src/utils/ManagedIdentityUtils.ts

bgavrilMS replied:
Then rename the variable as `expiresIn`?

`expiresOn` - name suggests this is a date
`expiresIn` - name suggestions this is a duration in seconds

When it stores tokens in the cache, MSAL uses dates.

bgavrilMS reviewed (changes requested):
- public API needs a review and it needs to  include PM, as there seems to be confusion. Let's get a sing-off from @localden 
- some existing public API like AuthenticationResult must not be modified
- tests do not check cache keys
- more complex tests needed
- skipped tests must be fixed

Robbie-Microsoft committed:
Implemented GitHub Feedback and re-worked some core functionality

Robbie-Microsoft committed:
Implemented Feedback

Robbie-Microsoft committed:
Unit test improvements

Robbie-Microsoft committed:
minor improvements

Robbie-Microsoft committed:
Fixed broken unit tests. ManagedIdentityClient is no longer static.

Robbie-Microsoft committed:
unit tests, and small changes in core functionality to support them

bgavrilMS reviewed (approved):
Approved with comments about testing the cache keys, which you are working on

Robbie-Microsoft committed:
Implemented cache key verification unit test
